gr-soapy/lib/block_impl.cc

```diff
@@ -1157,25 +1244,43 @@ std::string block_impl::read_uart(const std::string& which, long timeout_us) con
 void block_impl::cmd_handler_frequency(pmt::pmt_t val, size_t channel)
 {
     if (!(val->is_number() && !val->is_complex())) {
-        GR_LOG_WARN(d_logger, "soapy: freq must be float/int");
+        GR_LOG_ERROR(d_logger, "soapy: freq must be float/int");
         return;
     }
     set_frequency(channel, pmt::to_double(val));
 }
 
 void block_impl::cmd_handler_gain(pmt::pmt_t val, size_t channel)
 {
-    if (!(val->is_number() && !val->is_complex())) {
-        GR_LOG_WARN(d_logger, "soapy: gain must be float/int");
+    // For compatibility, accept either a numeric value for setting the
+    // overall gain or a dict with a gain element and element-specific
+    // value.
+    if (!(val->is_number() && !val->is_complex()) && !val->is_dict()) {
+        GR_LOG_ERROR(d_logger, "soapy: gain must be float/int or a dict");
         return;
     }
-    set_gain(channel, pmt::to_double(val));
+
+    if (val->is_dict()) {
+        if (!pmt::dict_has_key(val, CMD_GAIN_KEY)) {
+            GR_LOG_ERROR(d_logger, "soapy: gain dict must contain key \"gain\"");
+            return;
+        }
+
+        // Accept no name, falls back to default argument ""
+        const auto name =
+            pmt::symbol_to_string(pmt::dict_ref(val, CMD_NAME_KEY, pmt::mp("")));
+        const auto gain = pmt::to_double(pmt::dict_ref(val, CMD_GAIN_KEY, 0));
```

Comment:
uhhh I think this casts `0` to a shmart pointer? do we really intend to do that?

Reply:
`dict_ref(const pmt_t&, const pmt_t&, const pmt_t&)`: I wouldn't know how to convert 0 to a `const pmt_t==const std::shared_ptr<pmt_base>`

Reply:
Yay implicit casts. That should be fixed over in gr-uhd as well, and who knows where else:

https://github.com/gnuradio/gnuradio/blob/master/gr-uhd/lib/usrp_block_impl.cc#L731

Reply:
thanks!